### PR TITLE
Issues warnings for legend handles without handlers

### LINKED ISSUE #19121: Handle and label not created for Text with label
### Bug report

**Bug summary**

Text accepts a `label` keyword argument but neither its handle nor its label is created and added to the legend.

**Code for reproduction**

```python
import matplotlib.pyplot as plt

x = [0, 10]
y = [0, 10]

fig = plt.figure()
ax = fig.add_subplot(1, 1, 1)

ax.plot(x, y, label="line")
ax.text(x=2, y=5, s="text", label="label")

ax.legend()

plt.show()
```

**Actual outcome**

t

**Expected outcome**

I expect a legend entry for the text.

**Matplotlib version**
  * Matplotlib version: 3.3.3

### PR DESCRIPTION
**Pull request by kdpenner**
## PR Summary

Closes #19121.

Text was not included in the Artists collected for a legend, so text created with a label had no legend entry. This PR issues warnings when:

1. Text with a label is found
2. a rectangular handle is substituted for a Patch that is not a Rectangle

## PR Checklist

- [x] Has pytest style unit tests (and `pytest` passes).
- [x] Is [Flake 8](https://flake8.pycqa.org/en/latest/) compliant (run `flake8` on changed files to check).
- [N/A] New features are documented, with examples if plot related.
- [N/A] Documentation is sphinx and numpydoc compliant (the docs should [build](https://matplotlib.org/devel/documenting_mpl.html#building-the-docs) without error).
- [x] Conforms to Matplotlib style conventions (install `flake8-docstrings` and run `flake8 --docstring-convention=all`).
- [N/A] New features have an entry in `doc/users/next_whats_new/` (follow instructions in README.rst there).
- [N/A] API changes documented in `doc/api/next_api_changes/` (follow instructions in README.rst there).

### CODE REVIEW

**Comment by kdpenner** [comment]
Ready for review. FYI I received no notification that the PR was marked as a draft and assigned a status of needs tests until I checked up on my PRs.

**Comment by jklymak** [comment]
Ooops apologies.  I thought folks get those notifications.  In the future I will be sure to also comment.

File: lib/matplotlib/legend.py [hunk 1]
```diff
@@ -1141,27 +1149,19 @@ def _get_legend_handles(axs, legend_handler_map=None):
...
         # support parasite axes:
         if hasattr(ax, 'parasites'):
             for axx in ax.parasites:
                 handles_original += [
                     *(a for a in axx._children
-                      if isinstance(a, (Line2D, Patch, Collection))),
+                      if isinstance(a, (Line2D, Patch, Collection, Text))),
                     *axx.containers]
 
-    handler_map = Legend.get_default_handler_map()
```
Comment (tacaswell) [inline lib/matplotlib/legend.py:1154]: Why did you remove this whole code block?  Without this here one of the inputs in now completely and silently ignored.

File: lib/matplotlib/legend.py [hunk 2]
```diff
@@ -763,28 +764,35 @@ def _init_legend_box(self, handles, labels, markerfirst=True):
...
-                                                         fontsize, handlebox))
-                handles_and_labels.append((handlebox, textbox))
+                if isinstance(orig_handle, Text):
+                    _api.warn_external(
+                            "Legend does not support handles for Text "
+                            "instances.\nSee: https://matplotlib.org/stable/"
+                            "tutorials/intermediate/legend_guide.html"
+                            "#implementing-a-custom-legend-handler")
+                    continue
+                else:
```
Comment (tacaswell) [inline lib/matplotlib/legend.py:774]: This is going to provide switch _every_ Artists that is not `Text` to fallback to `Rectangle`!  That seems a bit over-broad and I am not sure the `Rectangle` handler will be able to extract the information it needs from an arbitrary `Artist`.

File: lib/matplotlib/legend.py (hunk 1 above)
Comment (kdpenner) [reply lib/matplotlib/legend.py:1154]: Hm. I don't remember the call sequence well enough to know why I did it in the first place. But I have an example where I pass a `handler_map` dict in the `legend` call and this block doesn't update the handler map:
```python
ax.legend(handles, labels, handler_map={Text: HandlerText()})
```
so I might've deleted it because it did nothing.

File: lib/matplotlib/legend.py (hunk 2 above)
Comment (kdpenner) [reply lib/matplotlib/legend.py:774]: The only `Artists` that are iterated over by default are: `Line2D`, `Patch`, `Collection`, and `Text`.

`Line2D` and (some?) `Collection` are in the `_default_handler_map`.

`Text` cannot be put in `_default_handler_map` unless a handler is written for it, and the solution for the linked issue was scoped to issuing a warning.

Before this PR, any `Patch` mapped to a `HandlerPatch`, which was a `Rectangle`. I've made that behavior explicit.

**Comment by kdpenner** [comment]
I have to update the PR to accommodate the new `Wedge` artist.

File: lib/matplotlib/legend.py (hunk 2 above)
Comment (tacaswell) [reply lib/matplotlib/legend.py:774]: This is called from https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L493-L494 which are lightly processed at https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L401-L411 but otherwise come in from https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L294-L295 which is in turn called from https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/axes/_axes.py#L299-L305 which has the path https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L1210-L1211 and https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L1240-L1242 that means the user can (eventually) get arbitrary artists into `handles` here so we can not assume that the input types will be limited to what the auto-discovery code picks up.

File: lib/matplotlib/legend.py (hunk 1 above)
Comment (tacaswell) [reply lib/matplotlib/legend.py:1154]: That may be evidence that there is a bug someplace else that is making it so the additional handler map is not getting routed.

I suspect that the change we want to make in this function is to drop the filtering on L1144 all together, put this block of code back, filter out the things with `_nolegend_` as the label, warn on things that do not have a handler (but drop them on the floor), and then restore the behavior of only yielding things that have handlers (that way we can give a different warning for auto-discovered things rather than explicitly passed things).

If we do this then I do not think we need any changes to `_init_legend_box` ?

File: lib/matplotlib/legend.py [hunk 3]
```diff
@@ -1141,27 +1149,19 @@ def _get_legend_handles(axs, legend_handler_map=None):
...
-    if legend_handler_map is not None:
-        handler_map = handler_map.copy()
-        handler_map.update(legend_handler_map)
-
-    has_handler = Legend.get_legend_handler
-
     for handle in handles_original:
         label = handle.get_label()
-        if label != '_nolegend_' and has_handler(handler_map, handle):
+        if label not in ['_nolegend_', '']:
```
Comment (tacaswell) [inline lib/matplotlib/legend.py:1164]: We currently do the filtering to "non empty labels" in https://github.com/matplotlib/matplotlib/blob/3e32cc7e88d6d583942335dff4c83b4b7dd13ae9/lib/matplotlib/legend.py#L1139-L1152 it is probably better if we leave that filtering step there and continue to only do "artist noped out of being in the legend" here.

File: lib/matplotlib/legend.py (hunk 2 above)
Comment (kdpenner) [reply lib/matplotlib/legend.py:774]: But the arbitrary artist needs a handler for it to be rendered in the legend; can a user pass an arbitrary artist in this way without one? So `handler` would not be `None` here:

https://github.com/matplotlib/matplotlib/blob/7d43704830e92059e40dcdf09561ec0f50bac37a/lib/matplotlib/legend.py#L765

and it wouldn't need to fall back to a `Rectangle`.

Comment (tacaswell) [reply lib/matplotlib/legend.py:774]: However if the user _does_make a mistake and pass in an artist that does not have a handler entry we should not then fallback to `Rectangle`, we should maintain the current behavior of warning and passing on the handle.

**Review by tacaswell — DISMISSED** [review]
See my in-line comments.  This PR is currently making more extensive changes to behavior than is required to add a warning when a label is added to a `Text` object.

**Comment by jklymak** [comment]
ping @kdpenner, this is draft again ;-)

**Comment by kdpenner** [comment]
@jklymak thanks for the ping, I got busy with work and let this languish.

@tacaswell this is ready for re-review.

Below are a bunch of usage patterns. ax2 and ax6 are important, as they are workarounds from before this PR, which changes their behavior. ax2 results in a nuisance text warning; previously no warning was issued. ax6 results in an additional entry in the legend.

```python
import matplotlib.pyplot as plt
from matplotlib.text import Text

class HandlerText:
    def legend_artist(self, legend, orig_handle, fontsize, handlebox):
        x0, y0 = handlebox.xdescent, handlebox.ydescent
        handle_text = Text(x=x0, y=y0, text=orig_handle.get_text())
        handlebox.add_artist(handle_text)
        return handle_text

fig = plt.figure()


print("ax1")
ax1 = fig.add_subplot(3, 3, 1)
# From original issue: now text warning is issued
ax1.text(x=2, y=5, s="text", label="label")
ax1.legend()


print("ax2")
ax2 = fig.add_subplot(3, 3, 2)
# With solution from before PR, nuisance text warning is issued
annotation = ax2.text(x=0, y=0, s="text", label="label")
handles, labels = ax2.get_legend_handles_labels()
handles.append(annotation)
labels.append(annotation.get_label())
ax2.legend(handles, labels, handler_map={Text: HandlerText()})


print("ax3")
ax3 = fig.add_subplot(3, 3, 3)
# Text warning is not issued
ax3.text(x=0, y=0, s="text", label="label")
ax3.legend(handler_map={Text: HandlerText()})


print("ax4")
ax4 = fig.add_subplot(3, 3, 4)
# Artists are found but their handlers are not added to the legend because
# they're not in the handler map passed in ax.legend()
ax4.text(x=0, y=0, s="text", label="label")
handles, labels = ax4.get_legend_handles_labels(
                                    legend_handler_map={Text: HandlerText()})
ax4.legend(handles, labels)


print("ax5")
ax5 = fig.add_subplot(3, 3, 5)
# Artists are found and added
ax5.text(x=0, y=0, s="text", label="label")
handles, labels = ax5.get_legend_handles_labels(
                                    legend_handler_map={Text: HandlerText()})
ax5.legend(handles, labels, handler_map={Text: HandlerText()})


print("ax6")
ax6 = fig.add_subplot(3, 3, 6)
from matplotlib.legend import Legend
Legend.update_default_handler_map({Text: HandlerText()})
# two legend entries
annotation = ax6.text(x=0, y=0, s="text", label="label")
handles, labels = ax6.get_legend_handles_labels()
handles.append(annotation)
labels.append(annotation.get_label())
ax6.legend(handles, labels)


print("ax7")
ax7 = fig.add_subplot(3, 3, 7)
# after handler map update, just works
ax7.text(x=0, y=0, s="text", label="label")
ax7.legend()

plt.show()
```

**Comment by jklymak** [comment]
This has a failing test: lib/matplotlib/tests/test_axes.py::test_pie_nolabel_but_legend[png]

**Comment by kdpenner** [comment]
That test fails because it doesn't expect the warning.

`plt.pie` creates Text artists with empty strings for the labels. `_get_legend_handles_labels` is called, which calls `_get_legend_handles`, which raises the warning. I could filter out the Text artists with `label not in ['_nolegend_', '']` in `_get_legend_handles`, but @tacaswell recommended leaving the filtering step to `_get_legend_handles_labels`, where artists with labels starting with '_' are dropped.

So there are three options:

1. use `label not in ['_nolegend_', ''] and not has_handler` in `_get_legend_handles`

https://github.com/matplotlib/matplotlib/blob/e82af34c4cf784718fe26725d81d3a0be50d2a2b/lib/matplotlib/legend.py#L1124

2. create Text artists with `_nolegend_` labels and use `label != "_nolegend_" and not has_handler` in `_get_legend_handles`
3. modify the test to capture the warning

**Comment by kdpenner** [comment]
Now that I've thought about this more option 1 is the only one for which warnings won't be raised for Text created without an intention of appearing in the legend.

**Comment by jklymak** [comment]
Popping to draft until you sort the above out.  I think @timhoffm was the proponent of this warning.

**Comment by kdpenner** [comment]
@jklymak what do I need to sort out? Thought I’d leave those options and discussion for the reviewer.

Also I think the needs test tag can be removed, as there is now a test.

**Comment by jklymak** [comment]
At the least it must pass the test suite.

**Comment by kdpenner** [comment]
do I need to mark the PR as ready for review for the workflows to run?

**Review by jklymak — APPROVED** [review]
This seems fine to me - I'm not 100% in favour of warnings like this; surely most folks understand that a text object doesn't need a legend entry.  But it may be helpful for other artists...

**Comment by tacaswell** [comment]
Thank you @kdpenner and congratulations on your first merged matplotlib PR 🎉 

Thank you for persevering with us through what was a long process!

**Comment by tacaswell** [comment]
Hopefully we will hear from you agian

**Comment by kdpenner** [comment]
Thanks!
